### PR TITLE
EM_ASM rework - doubles and varargs

### PR DESCRIPTION
**Pull request by jgravelle-google**
Building on https://github.com/kripken/emscripten/pull/6052

The plan is:
1) Convert all EM_ASM calls to taking only doubles. When numbers get to JS they get converted anyway, so if we do them C-side we haven't changed behavior.
2) Change EM_ASM signatures to `int emscripten_asm_const_int(const char* code, int num_args, ...)`. This will require changing the JS glue to fetch arguments out of the vararg memory address.
3) Simplify JS glue, no need to have separate tables by signature. Simplify binaryen asm const walker pass. Simplify metadata emitted if possible?

So far this is just the C macro side of casting arguments to always be doubles. This is technically landable as-is, but doesn't buy us any complexity reductions yet so probably no point in merging yet. Sending the PR out now to get feedback on the general plan.

### CODE REVIEW

**Comment by dschuff** [comment]
Can we use this for C++ too? Would it be simpler to do it for both than using templates as we are now?

**Comment by jgravelle-google** [comment]
We can and we should. Currently the change is in the macros, so the C++ calls are all-double as well, but via templates, whereas the C calls are all-double, but via prototypeless declarations. Next step is to unify them to both use varargs. Interestingly, I predict this will work without any LLVM or Binaryen changes, just this header and the associated JS.

File: system/include/emscripten/em_asm.h [hunk 1]
```diff
@@ -1,13 +1,72 @@
...
+// natively, so we can convert all C types to double before the call. By
+// converting all arguments to double, we can store them in a vararg buffer,
+// and we only need to pass the number of args in order to read them in JS.
+
+// When converting arguments to double, we need to:
+//   - upcast floats to doubles, so we have consistent bit widths
+//   - convert ints to doubles (via f64.convert_s/i32 and such)
+//   - cast pointers to ints, and then convert (C disallows direct pointer to
+//     double conversions)
+// We can use the _Generic clang extension / C11 feature to decide at compile
```
Comment (dschuff) [inline system/include/emscripten/em_asm.h]: Let's use the more-correct (and more importantly, more-searchable) term "generic selection" here. And maybe calling it a C11 feature first (which as an extension clang supports everywhere) makes it sound less dodgy :)

File: system/include/emscripten/em_asm.h [hunk 2]
```diff
@@ -1,13 +1,72 @@
...
+// time what operations to apply. Unfortunately each branch needs to compile for
+// all possible types, even though only one is selected. Fortunately, we can use
+// a function inside the _Generic expression, and use that do do our conversion.
+// All pointer types should go through the default: case.
+#define _EA_CAST(x) _Generic((x), \
```
Comment (kripken) [inline system/include/emscripten/em_asm.h]: maybe `EA => EM_ASM`? still pretty short, and more obvious what it does.

**Comment by kripken** [comment]
A long-term concern here is that i64s may arrive in JavaScript, at which point using doubles for everything won't work.

**Comment by jgravelle-google** [comment]
We could define EM_ASM to not work with i64s, which is something of a future-regression (or a non-extension point). It will be something of an odd asymmetry that we can export i64s to JS, but not via EM_ASM.

@dschuff suggested that when it comes to that, we can have a separate API, say EM_ASM_SIG, where the user provides the type signature of the arguments, and we can support it that way.
I think that not being able to support i64s without an API change is the least-breaking of the possible breaking changes to EM_ASM I've come across so far. The changes I've seen/considered include:
1) we drop support for EM_ASM from C entirely (C++ variadic templates are fine, obsolescent C prototypeless function declarations are not)
2) we add EM_ASM-specific magic to upstream LLVM and go back to asmjs-style vararg calling convention. Asm.js gets away with this because the place where it does its EM_ASM magic is early enough that it still has type information into the vararg buffer. This would be more complexity (we'd need to plumb metadata through the pipeline in some way, and/or more stages of the pipeline would need to be EM_ASM-aware), as well as code bloat for other language frontends that may not care.
3) change the API so we need to add a signature to every callsite of EM_ASM, breaking all user code that calls it

So I'm kind of pretty ok with that. Another thing to consider is: if you want to export i64s as BigInts, is EM_ASM the correct tool for the job? Does it need to be? We have other ways to interop between C and JS, maybe having a hole here is fine.

**Comment by kripken** [comment]
Good points. I think I still don't fully understand the problem, though - how does `asm()` (normal inline assembly) avoid the problems that are forcing us to change things?

**Comment by jgravelle-google** [comment]
I don't know enough about the `asm` keyword to say specifically for sure, but I think the key piece is it's recognized by the frontend directly, and part of the C language spec. It's also able to actually paste the asm block inline, vs. having to call out across a runtime boundary. It also puts more burden on the user to do things properly "or else".

But I think our main difficulties here are 1) we need our asm to be valid C, 2) we need to autogenerate a signature that we can use on the JS<->wasm boundary, 3) we aren't part of the C spec, but want to add a feature in what is roughly language-extension space, without modifying clang to do it.

**Comment by kripken** [comment]
I see, thanks. I wonder if since `asm()` has the type info because of the compiler's special handling, we could use that. That's what emscripten used to do a long time ago, and the only reason we stopped is because of escaping. Not sure if there's a solution to the escaping issues...

Anyhow, thinking more on this here, another option might be to, instead of using doubles for everything, to use i64s to represent the bits? That might add a little overhead (2 32 bit params per original param) but would be future proof. If that seems possible, we don't need to do it now, it could be a future way to support i64s in EM_ASM, assuming we can do it without user-noticeable effects.

**Comment by jgravelle-google** [comment]
I'd think that we'd want to emit wasm assembly from `asm`, inline with the compiled C code, instead of using JS as our asm, regardless of escaping issues.

And we could use i64s and bitcasting, but then we're back in the realm of needing to know what the signature WAS in order to read it correctly. I.e., we need to decide whether to read these 8 bytes as an i64 or an f64 in JS. f64s serve as a least-common-denominator that can represent i32, f32, and f64 *values*, without needing to know what type we started with. And it's the not needing to know that's beneficial here.

**Comment by kripken** [comment]
Hmm, right, the i64 bits approach can't work without JS knowing the type...

Would we be able to issue a compiler warning on people trying to use i64s with EM_ASM? At least we could point people to alternatives if so.

**Comment by jgravelle-google** [comment]
Semi-aside: https://github.com/kripken/emscripten/pull/6141/commits/c45404d45d18cc5f20173b0bd759fa7b85a8f681 makes EM_ASM work in C with LLD. This didn't work before because the prototypeless trick only worked in s2wasm because we were emitting text. Emitting binaries (e.g. object files) means we confound binaryen's wasm parser because the import has a mismatched type from how we call it, so binaryen adds drops to the seemingly-unused args, which causes problems.

**Comment by jgravelle-google** [comment]
Not sure if we'd be able to make it a compile-time failure, but we could definitely add a runtime error (or warning) here: https://github.com/kripken/emscripten/pull/6141/files#diff-5564d92b17fdcd584b79e85339491568R29

Could maybe make a separate `_ea_cast_long` function with a `noinline` attribute, then check if the compiled wasm has that function in it and warn at compile time if so?

File: system/include/emscripten/em_asm.h [hunk 3]
```diff
@@ -1,35 +1,84 @@
...
+  return x;
+}
+inline double _ea_cast_int(long x) {
+  return (double)x;
+}
+inline double _ea_cast_uint(unsigned long x) {
+  return (double)x;
+}
+inline double _ea_cast_ptr(const void* x) {
+  return (double)(long)x;
```
Comment (kripken) [inline system/include/emscripten/em_asm.h]: could `long` be `size_t`?

**Comment by kripken** [comment]
I see, sgtm then.

**Comment by jgravelle-google** [comment]
Actually... instead of counting the number of arguments, we could probably use `_Generic` to build a string like so:
```c
#define SIG(x) _Generic((x), \
    float: "f", \
    double: "d", \
    int: "i", \
    unsigned: "i", \
    long: "j", \
    unsigned long: "j", \
    default: "i")
```
then pass the string to JS, and use that to determine how to read the vararg buffer. It would be more complex, but we could export i64s today if we wanted to (not sure what we'd do with them atm, convert to double maybe?)

...
Actually trying that out doesn't work - C won't string-paste the strings across `_Generic`s. i.e.
```c
int puts(char*);
int main() {
    int x;
    float y;
    long z;
    puts(SIG(x) SIG(y) SIG(z));
    return 0;
}
```
I'm still a bit confused by `_Generic`. I guess it's compile-time, but not part of the preprocessor.
We could build our own string array, though not immediately as a literal:
```c
int main() {
    int x;
    float y;
    long z;
    char str[] = {SIG(x), SIG(y), SIG(z), 0};
    puts(str);
    return 0;
}
```
so we'd need to have some way to get a str in an expression-y context. Maybe a global str, and a vararg function to write the str and then return its pointer?

This is all starting to sound increasingly horrible but I mention it for completeness.

File: system/include/emscripten/em_asm.h (hunk 3 above)
Comment (jgravelle-google) [reply system/include/emscripten/em_asm.h]: `size_t` turns out to not be defined by that point. I think in principle it should be that, but `long` should conservatively work too?

Comment (kripken) [reply system/include/emscripten/em_asm.h]: I guess so, yeah.

**Comment by aheejin** [comment]
It looks like this broke these tests on Wasm backend:
- test_getFuncWrapper_sig_alias
- test_em_asm_parameter_pack
- test_em_asm_unicode
- test_dyncall
- test_dyncall_specific

Could you check? Here is the [log](https://logs.chromium.org/v/?s=chromium%2Fbb%2Fclient.wasm.llvm%2Flinux%2F28517%2F%2B%2Frecipes%2Fsteps%2FExecute_emscripten_testsuite__emwasm_%2F0%2Fstdout).

**Pull request merged**